File: 0x08-python-more_classes/4-rectangle.py
#!/usr/bin/python3
'''This module defines a rectangle with private instance variables'''


class Rectangle:
    '''This class defines a rectangle'''
    def __init__(self, width=0, height=0):
        '''Instantiation with optional width and height'''
        self.height = height
        self.width = width

    def __repr__(self):
        '''__retr__'''
        return "Rectangle("+str(self.__width)+", "+str(self.__height)+")"

    def __str__(self):
        '''__str___'''
        shape = ""
        if (self.__height == 0 or self.__width == 0):
            return shape
        for row in range(self.__height):
            for col in range(self.__width):
                shape += "#"
            if (row != self.height - 1):
                shape += "\n"
        return shape

    @property
    def height(self):
        '''height Getter'''
        return self.__height

    @height.setter
    def height(self, value):
        '''height Setter'''
        if (isinstance(value, int)):
            if (value >= 0):
                self.__height = value
            else:
                raise ValueError('height must be >= 0')
        else:
            raise TypeError('height must be an integer')

    @property
    def width(self):
        '''width getter'''
        return self.__width

    @width.setter
    def width(self, value):
        '''width setter'''
        if (isinstance(value, int)):
            if (value >= 0):
                self.__width = value
            else:
                raise ValueError('width must be >= 0')
        else:
            raise TypeError('width must be an integer')

    def area(self):
        return (self.__height * self.__width)

    def perimeter(self):
        if (self.__height == 0 or self.__width == 0):
            return 0

        return((2 * self.__height) + (2 * self.__width))

File: 0x13-javascript_objects_scopes_closures/4-rectangle.js
#!/usr/bin/node
class Rectangle {
  constructor (w, h) {
    if (w > 0 && h > 0) {
      this.width = w;
      this.height = h;
    }
  }

  print () {
    let concat = '';
    for (let i = 0; i < this.height; i++) {
      for (let j = 0; j < this.width; j++) {
        concat = concat + 'X';
      }
      console.log(concat);
      concat = '';
    }
  }

  rotate () {
    const tempWidth = this.width;
    this.width = this.height;
    this.height = tempWidth;
  }

  double () {
    this.width = 2 * this.width;
    this.height = 2 * this.height;
  }
}
module.exports = Rectangle;
